### PR TITLE
Continuous attribute embedding

### LINKED ISSUE #99: Embed Continuous Attributes
## Problem to Solve

Presently embedding categorical attribute values is supported, but the same is not true of attributes representing continuous values.

A categorical attribute is one which can take on a value that represents a category, e.g. `fever, light-sensitivity` or `1, 2, 3`

A continuous attribute is non-categorical e.g. any double value in the range `0.0 - 1.0`, or any integer in the range zero or above.

## Current Workaround

None

## Proposed Solution

Implement embedding module(s) for the neural network

### PR DESCRIPTION
**Pull request by jmsfltchr**
## What is the goal of this PR?

Enable ingesting numerical attributes with continuous values.

The aim has been to add a continuous numerical attribute to the diagnosis example which adds no additional information. In this case, the model should be able to stably achieve the same performance as without this attribute. Empirically, this has been achieved, perhaps taking longer to converge (about 500 training iterations minimum, compared to a minimum of 250 iterations prior).

Closes #99 

## What are the changes implemented in this PR?

- Introduce a continuous numerical attribute into the diagnosis example (`severity`)
- Create a `ContinuousAttribute` model, consisting of an MLP followed by layer normalisation
- For debugging and monitoring purposes, add histograms at strategic points in the model, plus the relevant code to execute and store these summaries
- Normalisation of common (type) embeddings for improved stability, such that attribute and type embeddings have similar magnitudes
- Gradient clipping for a very significant improvement in model convergence and stability
- Dropout for continuous attribute MLP to combat overfitting
- Improved the flow of embedder construction

### CODE REVIEW

File: kglib/kgcn/pipeline/pipeline.py [hunk 1]
```diff
@@ -136,3 +124,57 @@ def make_blank_embedder():
 
     _, _, _, _, _, solveds_tr, solveds_ge = tr_info
     return ge_graphs, solveds_tr, solveds_ge
+
+
+def configure_embedders(node_types, attr_embedding_dim, categorical_attributes, continuous_attributes):
```
Comment (flyingsilverfin) [inline kglib/kgcn/pipeline/pipeline.py]: this is a bit awkward to read... maybe it can be split into separate file/class?

File: kglib/kgcn/pipeline/pipeline_test.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+#
+import unittest
+
+from kglib.kgcn.pipeline.pipeline import configure_embedders
+
+
+class TestConstructEmbedders(unittest.TestCase):
```
Comment (flyingsilverfin) [inline kglib/kgcn/pipeline/pipeline_test.py]: already being tested separatedly!

File: kglib/kgcn/learn/learn.py [hunk 3]
```diff
@@ -70,20 +70,36 @@ def __call__(self,
...
         loss_ops_ge = loss_ops_preexisting_no_penalty(target_ph, output_ops_ge)
         loss_op_ge = loss_ops_ge[-1]  # Loss from final processing step.
+        tf.summary.scalar('loss_op_ge', loss_op_ge)
 
         # Optimizer
         optimizer = tf.train.AdamOptimizer(learning_rate)
-        step_op = optimizer.minimize(loss_op_tr)
+        gradients, variables = zip(*optimizer.compute_gradients(loss_op_tr))
+
+        for grad, var in zip(gradients, variables):
```
[the remark this answers is not in the document]
Comment (jmsfltchr) [reply kglib/kgcn/learn/learn.py]: Yes and also functional - `optimiser.minimize(loss)` computes the gradients and applies them. In order to intercept the gradients to do anything with them you have to first compute them and then apply them manually.

Here we do this to:
1. Visualise the gradients in TensorBoard
2. Apply gradient clipping as mentioned in the description

File: kglib/kgcn/pipeline/pipeline_test.py (hunk 2 above)
Comment (jmsfltchr) [reply kglib/kgcn/pipeline/pipeline_test.py]: You mean `configure_embedders` is? Whereabouts? 

The name of the test is incorrect though, should be `TestConfigureEmbedders`

File: kglib/kgcn/pipeline/pipeline.py (hunk 1 above)
Comment (jmsfltchr) [reply kglib/kgcn/pipeline/pipeline.py]: Agreed, it's generally awkward and needs more architectural work, including how to expose this configurability to the user. Intending to address this in a separate PR as it's a big task

Comment (jmsfltchr) [reply kglib/kgcn/pipeline/pipeline.py]: Added #103
